server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java

```diff
@@ -12,69 +12,71 @@
         // See https://github.com/elastic/elasticsearch/pull/19856.
-        IndexReader.CacheHelper cacheHelper = reader.getReaderCacheHelper();
-        CloseableThreadLocal<PerThreadIDVersionAndSeqNoLookup[]> ctl = lookupStates.get(cacheHelper.getKey());
-        if (ctl == null) {
-            // First time we are seeing this reader's core; make a new CTL:
-            ctl = new CloseableThreadLocal<>();
-            CloseableThreadLocal<PerThreadIDVersionAndSeqNoLookup[]> other = lookupStates.putIfAbsent(cacheHelper.getKey(), ctl);
-            if (other == null) {
-                // Our CTL won, we must remove it when the reader is closed:
-                cacheHelper.addClosedListener(removeLookupState);
-            } else {
-                // Another thread beat us to it: just use their CTL:
-                ctl = other;
-            }
-        }
 
-        PerThreadIDVersionAndSeqNoLookup[] lookupState = ctl.get();
-        if (lookupState == null) {
-            lookupState = new PerThreadIDVersionAndSeqNoLookup[reader.leaves().size()];
-            for (LeafReaderContext leaf : reader.leaves()) {
-                lookupState[leaf.ord] = new PerThreadIDVersionAndSeqNoLookup(leaf.reader(), uidField);
-            }
-            ctl.set(lookupState);
-        }
+        private final Map<IndexReader.CacheKey, PerThreadIDVersionAndSeqNoLookup[]> lookupsByReader = ConcurrentCollections
+            .newConcurrentMap();
 
-        if (lookupState.length != reader.leaves().size()) {
-            throw new AssertionError("Mismatched numbers of leaves: " + lookupState.length + " != " + reader.leaves().size());
+        private void removeLookup(IndexReader.CacheKey cacheKey) {
+            lookupsByReader.remove(cacheKey); // not called on the owning thread, hence the need for a CHM.
         }
 
-        if (lookupState.length > 0 && Objects.equals(lookupState[0].uidField, uidField) == false) {
-            throw new AssertionError(
-                "Index does not consistently use the same uid field: [" + uidField + "] != [" + lookupState[0].uidField + "]"
+        private PerThreadIDVersionAndSeqNoLookup[] getLookupState(
+            IndexReader.CacheHelper cacheHelper,
+            List<LeafReaderContext> leaves,
+            String uidField
+        ) throws IOException {
+            assert leaves.isEmpty() == false;
+            final var lookupState = lookupsByReader.computeIfAbsent(
+                cacheHelper.getKey(),
+                ignored -> new PerThreadIDVersionAndSeqNoLookup[leaves.size()]
             );
+            if (lookupState[0] == null) {
+                for (final var leaf : leaves) {
+                    assert lookupState[leaf.ord] == null;
+                    lookupState[leaf.ord] = new PerThreadIDVersionAndSeqNoLookup(leaf.reader(), uidField);
+                }
+                cacheHelper.addClosedListener(this::removeLookup);
```

Comment:
NB this may have performance impact since it means on close (i.e. refresh) we remove the entry from the map of every thread that touched it. Previously we'd have done most of that cleanup later on during GC.